### PR TITLE
Adding contains, intersects and matches operators

### PR DESCRIPTION
**Pull request by inmysocks**
Also added the documentation

### CODE REVIEW

File: core/modules/filters/contains.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/*\
+title: $:/core/modules/filters/contains.js
+type: application/javascript
+module-type: filteroperator
+
+Filter operator for comparing the contents of an input tiddlers field and the
+operand.
+
+This filter returns the input title iff the operand is a subset of the given
+field of the input.
```
Comment (pmario) [inline core/modules/filters/contains.js]: iff .. typo?

File: core/modules/filters/contains.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+MyList: 1 2
+filter: [[tiddler1]][[tiddler2]]+[contains<MyList>]
+
+tiddler1 list field: 1 2 foo hello
+
+tiddler1 is returned because the list field contains both 1 and 2
+
+tiddler2 list field: 1 foo hello
+
+tiddler2 is not returned because the list field doesn't contain 2
```
Comment (pmario) [inline core/modules/filters/contains.js]: does it only work with list fields?

File: core/modules/filters/contains.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+Example 2:
+
+MyList: foo bar
+filter: [[tiddler1]][[tiddler2]]+[contains:tags<MyList>]
```
Comment (pmario) [inline core/modules/filters/contains.js]: ok. ... It works with every field. right?

File: core/modules/filters/contains.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+Filter operator for comparing the contents of an input tiddlers field and the
+operand.
+
+This filter returns the input title iff the operand is a subset of the given
+field of the input.
+
+Example 1:
+
+MyList: 1 2
```
Comment (pmario) [inline core/modules/filters/contains.js]: what happens if I set: `MyList: 1 foo`

File: core/modules/filters/contains.js (hunk 1 above)
Comment (inmysocks) [reply core/modules/filters/contains.js]: no, iff as in 'if and only if', should I avoid using language like that in the documentation?

File: core/modules/filters/contains.js (hunk 4 above)
Comment (inmysocks) [reply core/modules/filters/contains.js]: Than it will return the tiddler because the tiddler also has foo in its field.

File: core/modules/filters/contains.js (hunk 2 above)
Comment (inmysocks) [reply core/modules/filters/contains.js]: It treats the field as a list, but since there is nothing to mark a field as a list field it works for any field. I used it on titles and it worked splitting the titles into separate items as expected.

File: core/modules/filters/contains.js (hunk 3 above)
Comment (inmysocks) [reply core/modules/filters/contains.js]: yes, it tries to parse it as a list so that may give strange behaviour, but that is the same as many of the list opts operators.

File: core/modules/filters/contains.js (hunk 1 above)
Comment (pmario) [reply core/modules/filters/contains.js]: If it isn't proper english ... yes

**Comment by diego898** [comment]
This is great. Thanks @inmysocks for making this!

**Comment by tobibeer** [comment]
Interesting, so [contains](http://tobibeer.github.io/tw5-plugins/#contains), clearly has some important use cases to more than just myself. ;-)

File: core/modules/filters/contains.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+exports.contains = function(source,operator,options) {
+	var results = [];
+  let invert = operator.prefix === "!";
+  // Default to using the list field
+  var field = operator.suffix || 'list';
+  source (function(tiddler,title) {
+    // Make sure the tiddler exists and has the field
+    if (tiddler && tiddler.fields[field]) {
+      // Make field into a set (array then set, but in one step)
+      let inputSet = new Set($tw.utils.parseStringArray(tiddler.fields[field]));
```
Comment (tobibeer) [inline core/modules/filters/contains.js]: What if I wanted to compare the actual titles in the input and not any of their fields?

File: core/modules/filters/contains.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      let operandSet = new Set($tw.utils.parseStringArray(operator.operand));
+      // Do the set intersection
+      let intersection = new Set([...inputSet].filter(x => operandSet.has(x)));
+      // Subtract the intersection from the operandSet
+      // This only works as a test of equality because we know that the
+      // intersection must be a subset of the operandSet. This does not work in
+      // general for set equality!
+      // Test if the difference is empty, if so the sets are equal
+      let pass = [...operandSet].filter(x => !intersection.has(x)).length === 0;
+      if (!pass && invert) {
```
Comment (tobibeer) [inline core/modules/filters/contains.js]: The general strategy / prefereance by @Jermolene is to not do logical lifting in the source function for performance reasons.

In other words, you should have X source functions, one for each usecase. Here it would be one for inversion and one for non-inverted.

File: core/modules/filters/intersects.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  var invert = operator.prefix === "!";
+  // Default to using the list field
+  var field = operator.suffix || 'list';
+  source (function(tiddler,title) {
+    // Make sure the tiddler exists and has the field
+    if (tiddler && tiddler.fields[field]) {
+      // Make field into a set (array then set, but in one step)
+      let inputSet = new Set($tw.utils.parseStringArray(tiddler.fields[field]));
+      // Make operator into a set
+      let operandSet = new Set($tw.utils.parseStringArray(operator.operand));
```
Comment (tobibeer) [inline core/modules/filters/intersects.js]: dito, outside

File: core/modules/filters/contains.js (hunk 1 above)
Comment (tobibeer) [reply core/modules/filters/contains.js]: no documentation in the modules, pls ...and no dev / tech lingo (iff)

File: core/modules/filters/intersects.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+(function(){
+
+/*jslint node: true, browser: true */
+/*global $tw: false */
+"use strict";
+
+/*
+Export our filter function
+*/
+exports.intersects = function(source,operator,options) {
```
Comment (tobibeer) [inline core/modules/filters/intersects.js]: What's the difference to contains? This PR really needs a demo to play with!

File: core/modules/filters/intersects.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    if (tiddler && tiddler.fields[field]) {
+      // Make field into a set (array then set, but in one step)
+      let inputSet = new Set($tw.utils.parseStringArray(tiddler.fields[field]));
+      // Make operator into a set
+      let operandSet = new Set($tw.utils.parseStringArray(operator.operand));
+      // Do the set intersection
+      let intersection = [...inputSet].filter(x => operandSet.has(x));
+      // Test if the intersection is empty
+      let pass = intersection.length !== 0;
+      if (!pass && invert) {
```
Comment (tobibeer) [inline core/modules/filters/intersects.js]: dito, extract into two source functions

File: core/modules/filters/matches.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  var invert = operator.prefix === "!";
+  // Default to using the list field
+  var field = operator.suffix || 'list';
+  source (function(tiddler,title) {
+    // Make sure the tiddler exists and has the field
+    if (tiddler && tiddler.fields[field]) {
+      // Make field into a set (array then set, but in one step)
+      let inputSet = new Set($tw.utils.parseStringArray(tiddler.fields[field]));
+      // Make operator into a set
+      let operandSet = new Set($tw.utils.parseStringArray(operator.operand));
```
Comment (tobibeer) [inline core/modules/filters/matches.js]: dito, outside

File: core/modules/filters/matches.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+      // Do the set intersection
+      let intersection = [...inputSet].filter(x => operandSet.has(x));
+      // Test if the intersection size is equal to the input size and the
+      // operand size. If so the input and operand are the same set.
+      let pass = intersection.length === inputSet.size && inputSet.size === operandSet.size;
+      if (!pass && invert) {
```
Comment (tobibeer) [inline core/modules/filters/matches.js]: dito, two source functions

File: editions/tw5.com/tiddlers/filters/examples/matches Operator (Examples).tid [hunk 12]
```diff
@@ -0,0 +1,8 @@
+title: matches Operator (Examples)
+tags: [[matches Operator]] [[Operator Examples]]
+
+Note: The first example can be done using the [[field operator|field Operator]], the second example shows one way this operater differs from field.
+The large conceptual difference between this operator and field is that for field you have to have the exact field contents, for this operator you just need to have list but the order of the items on the list isn't important.
```
Comment (tobibeer) [inline editions/tw5.com/tiddlers/filters/examples/matches Operator (Examples).tid]: when doing list filters, the order of the items should **never** be relevant!

File: core/modules/filters/matches.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
+/*\
+title: $:/core/modules/filters/matches.js
+type: application/javascript
+module-type: filteroperator
+
+Filter operator for comparing the contents of an input tiddlers field and the
```
Comment (tobibeer) [inline core/modules/filters/matches.js]: The difference to **contains** and **intersects** escapes me. Needs to be demoed and documented. And if there is none, it should simply be aliases.

File: editions/tw5.com/tiddlers/filters/examples/matches Operator (Examples).tid (hunk 12 above)
Comment (inmysocks) [reply editions/tw5.com/tiddlers/filters/examples/matches Operator (Examples).tid]: yes, for this operator order isn't relevant. That is why this works for matching the lists in two fields where the field operator wouldn't work.

File: core/modules/filters/intersects.js (hunk 8 above)
Comment (inmysocks) [reply core/modules/filters/intersects.js]: The demo is here http://ooktech.com/jed/ExampleWikis/SetFilters/#contains%20Operator:%5B%5Bcontains%20Operator%5D%5D%20%5B%5Bintersects%20Operator%5D%5D%20%5B%5Bmatches%20Operator%5D%5D

File: core/modules/filters/matches.js (hunk 13 above)
Comment (inmysocks) [reply core/modules/filters/matches.js]: contains means the input list contains everything on the operand list, intersects means that the input list and operand list share at least one element.

File: core/modules/filters/contains.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  var field = operator.suffix || 'list';
+  source (function(tiddler,title) {
+    // Make sure the tiddler exists and has the field
+    if (tiddler && tiddler.fields[field]) {
+      // Make field into a set (array then set, but in one step)
+      let inputSet = new Set($tw.utils.parseStringArray(tiddler.fields[field]));
+      // Make operator into a set
+      let operandSet = new Set($tw.utils.parseStringArray(operator.operand));
+      // Do the set intersection
+      let intersection = new Set([...inputSet].filter(x => operandSet.has(x)));
```
Comment (tobibeer) [inline core/modules/filters/contains.js]: Should we really be adopting ECMAscript 6 just yet, such as arrow functions?

**Comment by tobibeer** [comment]
I would rephrase the summary and docs:

**contains**

A filter operator returning all input titles whose list or list-like field contains all titles in the list of titles specified by the operand, i.e. where the titles in the operand is a subset of the titles listed in the list or list-like field of each tiddler in the list of input titles.

**intersect**

A filter operator returning all input titles whose list or list-like field contains some titles in the list of titles specified by the operand, i.e. where at least one title in the operand is contained in the titles listed in the list or list-like field of each tiddler in the list of input titles.

**matches**

A filter operator returning all input titles whose list or list-like field contains exactly those titles in the list of titles specified by the operand, i.e. where exactly those titles in the operand and no other are contained in the titles listed in the list or list-like field of each tiddler in the list of input titles.

----

The might beneit from a table that fills the blanks with those textbits that differ among the three.

As said before, What I am missing here is the ability to do contains, matches. and intersects directly against the input titles and not just their list or list-like fields.

I also think that it would be very beneficial to be able to say "stop at the first match you find". So that you don't need to validate a 100 tiddlers if all you wanted to know was if one such existed (which you could do using [limit[1]] afterwards, but which yields awful performance.

@Jermolene: In fact, I think it's almost sensible to suggest for all filters (where that makes sense) to allow for a numeric prefix that indicates "only take the X first", e.g.

`foo bar baz +1[matches<some-list>]`

=> check for foo, bar and baz if their list field matches exactly all titles in `<some-list>` and return only the first. So, if the list field of foo already contains all and only those titles in `<some-list>`, we should not proceed also checking bar and baz.

**Comment by inmysocks** [comment]
Checking against the input list of titles is a very different action than checking for a property of each input title. It would add a third class of filter operators (in addition to selection modifiers and selection constructors), it would be very useful to have but I don't want to overload the operators to do two completely different things, one of which is a completely new class of actions for filter operators.

**Comment by tobibeer** [comment]
Ok, so something like **inputcontains**, **inputmatches** or **inputintersects**...perhaps with output option as to what to return in which case.

**Comment by inmysocks** [comment]
It looks like a cleaner version of the contains operator was merged so I am going to close this and see if I can match that style with intersects and matches.
